[Infra] Add collapsible sections

jennypavlova opened the pull request:
Closes #175558

## Summary

This PR makes all asset details overview sections collapsible. All sections will be always open by default and the user should be able to expand/collapse a section.

 
<img width="300" alt="Screenshot 2024-02-01 at 12 04 24" src="https://github.com/elastic/kibana/assets/14139027/b91a190a-eaec-4031-8591-f11a3a04f201"> <img width="300" alt="Screenshot 2024-02-01 at 12 03 50" src="https://github.com/elastic/kibana/assets/14139027/9d8c20fc-3a27-481c-89ad-1fd7f0e6df47">

## Testing

1. Open Host view
2. Open the hosts flyout and go to the overview tab
   - All sections should be collapsible
   - All section buttons should be clickable

https://github.com/elastic/kibana/assets/14139027/27a6df67-c560-42d6-9166-a8f45880cf8e

3. Open the asset details page and go to the overview tab
   - All sections should be collapsible
   - All section buttons should be clickable

https://github.com/elastic/kibana/assets/14139027/f7225ddd-c2a3-4a7b-b520-22acb41e9080

jennypavlova committed:
Add collapsible sections in the asset details overview tab

jennypavlova committed:
Merge branch 'main' of github.com:elastic/kibana into 175558-infra-add-collapsible-sections-in-the-overview-tab

jennypavlova committed:
Update section id, add closed section content and no alerts case

jennypavlova committed:
Add tests and fix test ids

jennypavlova committed:
Change data test subjs

jennypavlova committed:
Try to use alerts count only in the alert widget and close section if there are no alerts

jennypavlova committed:
Update ids and alerts closed content

jennypavlova committed:
Refactoring: remove alerts customizations

jennypavlova committed:
Merge branch 'main' into 175558-infra-add-collapsible-sections

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/links/link_to_alerts.tsx:
```diff
@@ -9,7 +9,7 @@ import { FormattedMessage } from '@kbn/i18n-react';
 
 export interface LinkToAlertsRuleProps {
-  onClick?: () => void;
+  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
```
I think it'd be better if you handled the `e.stopPropagation();` directly in this component instead of passing up the element. We guarantee the same behaviour for all clients then. WDYT?

```js
onClick={(e) => {
   e.stopPropagation();
   onClick();
}}
```

jennypavlova replied:
Here we need this only because inside the accordion header (in the previous case or any other cases this is not needed) - [this example](https://eui.elastic.co/#/layout/accordion#interactive-content-in-the-trigger). Also `onClick` is `optional` parameter so [here](https://github.com/elastic/kibana/blob/da34e181cd8bdde7bc4a58c08f5bc9b4ef715360/x-pack/plugins/infra/public/components/asset_details/hooks/use_page_header.tsx#L98) in the header, for example, we don't need it.

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/common/popover.tsx:
```diff
@@ -27,7 +27,13 @@ export const Popover = ({
       button={
-        <button onClick={togglePopover} data-test-subj={props['data-test-subj']}>
+        <button
+          onClick={(e) => {
+            e.stopPropagation();
+            togglePopover();
+          }}
```
Following the comment above, you could go back to the way it was:

```suggestion
          onClick={togglePopover}
```

cauemarcondes replied:
ah my bad, I just thought it was using the `<LinkToAlertsRule` component

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/alerts.tsx:
```diff
@@ -47,27 +48,39 @@ export const AlertsSummaryContent = ({
-          <LinkToAlertsPage
-            assetName={assetName}
-            queryField={`${assetType}.name`}
-            dateRange={dateRange}
+          <LinkToAlertsRule
+            onClick={(e) => {
+              e.stopPropagation();
+              toggleAlertFlyout();
+            }}
           />
         </EuiFlexItem>
-      </EuiFlexGroup>
-      <EuiSpacer size="s" />
-      <MemoAlertSummaryWidget alertsQuery={alertsEsQueryByStatus} dateRange={dateRange} />
+      )}
+      <EuiFlexItem grow={false}>
+        <LinkToAlertsPage
+          assetName={assetName}
+          queryField={`${assetType}.name`}
+          dateRange={dateRange}
+        />
+      </EuiFlexItem>
+    </EuiFlexGroup>
+  );
```
I think you can move this component out of the `AlertsSummaryContent`, right?

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/metrics/metrics_section.tsx:
```diff
@@ -49,7 +49,7 @@ export const MetricsSection = ({ assetName, metricsDataView, logsDataView, dateR
   return (
     <EuiFlexGroup direction="column" gutterSize="s">
-      <Section title={MetricsSectionTitle}>
+      <Section title={MetricsSectionTitle} shouldCollapse={true}>
```
```suggestion
      <Section title={MetricsSectionTitle} collapsible>
```

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/metrics/metrics_section.tsx:
```diff
@@ -107,13 +107,14 @@ export const MetricsSectionCompact = ({
 const Section = ({
   title,
   dependsOn = [],
+  shouldCollapse = false,
```
I think you could probably name this: 

```suggestion
  collapsible = false,
```

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/section/collapsible_section.tsx:
```diff
@@ -0,0 +1,82 @@
+  extraAction?: React.ReactNode;
+  dependsOn?: string[];
+  children: React.ReactNode;
+  shouldCollapse: boolean;
```
Same here:
```suggestion
  collapsible: boolean;
```

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/metrics/metrics_section.tsx:
```diff
@@ -124,11 +125,15 @@ const Section = ({
+    <CollapsibleSection
+      title={title}
+      shouldCollapse={shouldCollapse}
+      data-test-subj={
+        shouldCollapse ? 'infraAssetDetailsMetricsCollapsible' : 'infraAssetDetailsMetricsSection'
+      }
```
nit: 
```suggestion
      data-test-subj={
        `infraAssetDetailsMetrics${collapsible ? 'Collapsible': 'Section'}
      }
```

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/section/collapsible_section.tsx:
```diff
@@ -0,0 +1,82 @@
+    closedSectionContent && trigger === 'closed' ? (
+      <EuiFlexGroup gutterSize="m">
+        <EuiFlexItem grow={false}>
+          <>
+            <Title />
+          </>
```
Is this extra Fragment necessary?
```suggestion
            <Title />
```

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/section/collapsible_section.tsx:
```diff
@@ -0,0 +1,82 @@
+  ['data-test-subj']: string;
+  id: string;
+}) => {
+  const [trigger, setTrigger] = useState<SectionTriggerValue>('open');
```
I think you can use the type defined in the Eui component, can't you?
```suggestion
  const [trigger, setTrigger] = useState<EuiAccordionProps['forceState']>('open');
```

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/alerts.tsx:
```diff
@@ -47,27 +48,39 @@ export const AlertsSummaryContent = ({
-            queryField={`${assetType}.name`}
-            dateRange={dateRange}
+          <LinkToAlertsRule
+            onClick={(e) => {
+              e.stopPropagation();
+              toggleAlertFlyout();
+            }}
```
```suggestion
            onClick={toggleAlertFlyout}
```

jennypavlova replied:
https://github.com/elastic/kibana/pull/176048/files#r1474804714

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/metadata_summary/metadata_summary_list.tsx:
```diff
@@ -78,54 +80,56 @@ const MetadataSummaryListWrapper = ({
     showTab(ContentTabIds.METADATA);
   };
 
+  const ExtraActions = () => (
+    <EuiButtonEmpty
+      data-test-subj="infraAssetDetailsMetadataShowAllButton"
+      onClick={onClick}
+      size="xs"
+      flush="both"
+      iconSide="right"
+      iconType="sortRight"
+      key="metadata-link"
+    >
+      <FormattedMessage
+        id="xpack.infra.assetDetailsEmbeddable.metadataSummary.showAllMetadataButton"
+        defaultMessage="Show all"
+      />
+    </EuiButtonEmpty>
+  );
```
As this `ExtraActions` only returns a single button, I think you could remove it and define the button directly on the `CollapsibleSection` on line 107, WDYT?

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/metadata_summary/metadata_summary_list.tsx:
```diff
@@ -78,54 +80,56 @@ const MetadataSummaryListWrapper = ({
+          <MetadataExplanationMessage />
+          <EuiSpacer size="s" />
+          <EuiFlexGroup>
+            {visibleMetadata.map(
+              (metadataValue) =>
+                metadataValue && (
```
I could filter out the undefined values first and them only render the ones defined.
```suggestion
            {visibleMetadata.filter(metadataValue => metadataValue).map(
              (metadataValue) =>
```

jennypavlova committed:
CR changes

jennypavlova commented:
@cauemarcondes Thank you for reviewing the PR! I added comments related to the `onClick` suggestions and fixed the other parts mentioned, can you please take a look again when you have time?

jennypavlova committed:
Fix collapsible prop values

jennypavlova committed:
Merge branch 'main' into 175558-infra-add-collapsible-sections

cauemarcondes commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/alerts.tsx:
```diff
@@ -49,25 +50,35 @@ export const AlertsSummaryContent = ({
+              <EuiFlexItem grow={false}>
+                <LinkToAlertsRule
+                  onClick={(e) => {
+                    // e.stopPropagation();
```
Maybe you can remove this?

cauemarcondes reviewed (approved):
LGTM

jennypavlova committed:
Remove not needed onClick action
